perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
+		serveMetric.MetricMap[rpsPerCore] = float64(realRequests) / (profiler.GetCoolDownTime() - profiler.GetWarmUpTime())
+		if cores > vmNum {
+			serveMetric.MetricMap[rpsPerCore] /= float64(vmNum)
+		} else {
+			serveMetric.MetricMap[rpsPerCore] /= float64(cores)
+		}
+		result, err := profiler.GetResult()
+		profiledCores := profiler.GetCores()
+		log.Debugf("%d cores are recorded: %v", len(profiledCores), profiledCores)
+		require.NoError(t, err, "Stopping profiler returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f", eventName, value)
+			serveMetric.MetricMap[eventName] = value
+		}
+		log.Debugf("%s: %f", avgExecTime, serveMetric.MetricMap[avgExecTime])
+		log.Debugf("%s: %f", rpsPerCore, serveMetric.MetricMap[rpsPerCore])
+		profiler.PrintBottlenecks()
+
+		// if tail latency violates the contraints that it should be less than 5x service time,
+		// it returns the metric before tail latency violation.
+		if latencies.tailLatency > threshold {
+			require.NotNil(t, pmuMetric, "The tail latency of first round is larger than the constraint")
+			return pmuMetric.MetricMap
+		}
+
+		pmuMetric = serveMetric
 	}
 
-	done <- true
-	vmGroup.Wait()
+	return pmuMetric.MetricMap
 }
 
-///////////////////////////////////////////////////////////////////////////////
-////////////////////////// Auxialiary functions below /////////////////////////
-///////////////////////////////////////////////////////////////////////////////
-func serveVM(t *testing.T, start time.Time, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool) {
+// loadVMs load requests to VMs every second and records completed requests and exection time
+func loadVMs(t *testing.T, vmGroup *sync.WaitGroup, vmID, requestID int, imageName string,
+	isSyncOffload bool, isProfile *bool, execTime, realRequests *int64) {
 	defer vmGroup.Done()
 
+	vmIDString := strconv.Itoa(vmID)
+	log.Debugf("VM %s: requestID %d", vmIDString, requestID)
+	tStart := time.Now()
 	resp, _, err := funcPool.Serve(context.Background(), vmIDString, imageName, "replay")
-	require.NoError(t, err, "Function returned error")
-	require.Equal(t, resp.Payload, "Hello, replay_response!")
+	require.Equal(t, resp.IsColdStart, false)
+	if err == nil {
+		if resp.Payload != "Hello, replay_response!" {
+			log.Debugf("Function returned invalid: %s", resp.Payload)
+		}
+		if *isProfile {
+			atomic.AddInt64(realRequests, 1)
+			atomic.AddInt64(execTime, time.Since(tStart).Milliseconds())
+			log.Debugf("VM %s: requestID %d completed in %d milliseconds", vmIDString, requestID, time.Since(tStart).Milliseconds())
+		}
+	} else {
+		log.Debugf("VM %s: Function returned error %v", vmIDString, err)
+	}
+}
 
-	if *isColdStart {
-		require.Equal(t, resp.IsColdStart, true)
-		message, err := funcPool.RemoveInstance(vmIDString, imageName, isSyncOffload)
+// configureProfiler controls the time duration of profiling for loadVMs and profiler
```

Comment:
for the loadVMs function and the profiler

Reply:
Done